ffi/src/metrics_setup.rs

```diff
@@ -16,39 +17,44 @@ use chrono::{DateTime, Utc};
 use metrics::Key;
 use metrics_util::registry::{AtomicStorage, Registry};
 
-static INIT: Once = Once::new();
-
-pub(crate) fn setup_metrics(metrics_port: u16) {
-    INIT.call_once(|| {
-        let inner: TextRecorderInner = TextRecorderInner {
-            registry: Registry::atomic(),
-        };
-        let recorder = TextRecorder {
-            inner: Arc::new(inner),
-        };
-        metrics::set_global_recorder(recorder.clone()).expect("failed to set recorder");
-
-        Server::new(move |request| {
-            if request.method() == "GET" {
-                Response::builder()
-                    .status(StatusCode::OK)
-                    .header("Content-Type", "text/plain")
-                    .body(Body::from(recorder.stats()))
-                    .expect("failed to build response")
-            } else {
-                Response::builder()
-                    .status(StatusCode::METHOD_NOT_ALLOWED)
-                    .body(Body::from("Method not allowed"))
-                    .expect("failed to build response")
-            }
-        })
-        .bind((Ipv4Addr::LOCALHOST, metrics_port))
-        .bind((Ipv6Addr::LOCALHOST, metrics_port))
-        .with_global_timeout(Duration::from_secs(60 * 60))
-        .with_max_concurrent_connections(2)
-        .spawn()
-        .expect("failed to spawn server");
-    });
+pub(crate) fn setup_metrics(metrics_port: u16) -> Result<(), Box<dyn Error>> {
+    let inner: TextRecorderInner = TextRecorderInner {
+        registry: Registry::atomic(),
+    };
+    let recorder = TextRecorder {
+        inner: Arc::new(inner),
+    };
+
+    // Any caller should ensure that the metrics system is not already initialized
+    // by explicitly calling with metrics_port == 0.
+    // If multiple databases are running, they should use the same port.
+    if let Err(e) = metrics::set_global_recorder(recorder.clone()) {
+        if metrics_port == 0 {
```

Comment:
This must be done *before* calling `set_global_recorder`. Otherwise, if you open two databases, the first one doesn't want metrics and the second one does, it won't work correctly.